include/rocksdb/fs_env.h

```diff
@@ -0,0 +1,1323 @@
+
+#pragma once
+
+#include <stdint.h>
+#include <cstdarg>
+#include <functional>
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+#include "rocksdb/env.h"
+#include "rocksdb/io_status.h"
+#include "rocksdb/thread_status.h"
+
+namespace rocksdb {
+
+class FileLock;
+class FSDirectory;
+class FSMemoryMappedFileBuffer;
+class FSRandomAccessFile;
+class FSRandomRWFile;
+class FSSequentialFile;
+class FSWritableFile;
+class Logger;
+class Slice;
+struct DBOptions;
+struct ImmutableDBOptions;
+struct MutableDBOptions;
+class RateLimiter;
+
+// The FSEnv, FSSequentialFile, FSRandomAccessFile, FSWritableFile,
+// FSRandomRWFileclass, and FSDIrectory classes define the interface between
+// RocksDB and storage systems, such as Posix filesystems,
+// remote filesystems etc.
+// The interface allows for fine grained control of individual IO operations,
+// such as setting a timeout, prioritization, hints on data placement,
+// different handling based on type of IO etc.
+// This is accomplished by passing an instance of IOOptions to every
+// API call that can potentially perform IO. Additionally, each such API is
+// passed a pointer to a IODebugContext structure that can be used by the
+// storage system to include troubleshooting information. The return values
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FSEnv {
+ public:
+  struct FileAttributes {
+    // File name
+    std::string name;
+
+    // Size of file in bytes
+    uint64_t size_bytes;
+  };
+
+  // Priority of an IO request. This is a hint and does not guarantee any
+  // particular QoS.
+  // IO_LOW - Typically background reads/writes such as compaction/flush
+  // IO_HIGH - Typically user reads/synchronous WAL writes
+  enum class IOPriority : uint8_t {
+    IO_LOW,
+    IO_HIGH,
+    IO_TOTAL,
+  };
+
+  // Type of the data begin read/written. It can be passed down as a flag
+  // for the FSEnv implementation to optionally handle different types in
+  // different ways
+  enum class IOType : uint8_t {
+    kData,
+    kFilter,
+    kIndex,
+    kMetadata,
+    kWAL,
+    kManifest,
+    kLog,
+    kUnknown,
+    kInvalid,
+  };
+
+  // Per-request options that can be passed down to the FSEnv implementation.
+  // These are hints and are not necessarily guaranteed to be honored.
+  // More hints can be added here in the future to indicate things like
+  // storage media (HDD/SSD) to be used, replication level etc.
+  struct IOOptions {
+    // Timeout for the operation in milliseconds
+    std::chrono::milliseconds timeout;
+
+    // Priority - high or low
+    IOPriority prio;
+
+    // Type of data being read/written
+    IOType type;
+  };
+
+  // A structure to pass back some debugging information from the FSEnv
+  // implementation to RocksDB in case of an IO error
+  struct IODebugContext {
```

Comment:
I am not sure I understand the use of this struct.  Why would there be a different one for every API call?  Who is managing this class (it is usually passed in as a pointer -- can it be null? what is the lifetime, etc?)

Would this be better as something that is a member of the FSEnv?

Reply:
I envisioned this being similar to PerfContext or IOStatsContext, i.e it could keep debugging counters for a single IO operation. I think its definition will have to go through a couple of iterations before it becomes meaningful. I will update the comments to clarify allocation, lifetime etc.